volatility3/framework/plugins/windows/debugregisters.py

```diff
@@ -0,0 +1,219 @@
+
+from typing import Tuple, Optional, Generator, List, Dict
+
+from functools import partial
+
+from volatility3.framework import renderers, interfaces, exceptions
+from volatility3.framework.configuration import requirements
+from volatility3.framework.renderers import format_hints
+import volatility3.plugins.windows.pslist as pslist
+import volatility3.plugins.windows.threads as threads
+import volatility3.plugins.windows.pe_symbols as pe_symbols
+
+vollog = logging.getLogger(__name__)
+
+
+class DebugRegisters(interfaces.plugins.PluginInterface):
+    # version 2.6.0 adds support for scanning for 'Ethread' structures by pool tags
+    _required_framework_version = (2, 6, 0)
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List:
+        return [
+            requirements.ModuleRequirement(
+                name="kernel",
+                description="Windows kernel",
+                architectures=["Intel32", "Intel64"],
+            ),
+            requirements.VersionRequirement(
+                name="pslist", component=pslist.PsList, version=(2, 0, 0)
+            ),
+            requirements.VersionRequirement(
+                name="pe_symbols", component=pe_symbols.PESymbols, version=(1, 0, 0)
+            ),
+        ]
+
+    def _get_debug_info(
+        self, ethread: interfaces.objects.ObjectInterface
+    ) -> Optional[Tuple[interfaces.objects.ObjectInterface, int, int, int, int, int]]:
+        """
+        Gathers information related to the debug registers for the given thread
+        """
+        try:
+            dr7 = ethread.Tcb.TrapFrame.Dr7
+            state = ethread.Tcb.State
+        except exceptions.InvalidAddressException:
+            return None
+
+        # 0 = debug registers not active
+        # 4 = terminated
+        if dr7 == 0 or state == 4:
+            return None
+
+        try:
+            owner_proc = ethread.owning_process()
+        except (AttributeError, exceptions.InvalidAddressException):
+            return None
+
+        dr0 = ethread.Tcb.TrapFrame.Dr0
+        dr1 = ethread.Tcb.TrapFrame.Dr1
+        dr2 = ethread.Tcb.TrapFrame.Dr2
+        dr3 = ethread.Tcb.TrapFrame.Dr3
+
+        # bail if all are 0
+        if not (dr0 or dr1 or dr2 or dr3):
+            return None
+
+        return owner_proc, dr7, dr0, dr1, dr2, dr3
+
+    def _get_vads(
+        self,
+        vads_cache: Dict[int, List[Tuple[int, int, str]]],
+        owner_proc: interfaces.objects.ObjectInterface,
+    ) -> Optional[List[Tuple[int, int, str]]]:
+        if owner_proc.vol.offset in vads_cache:
+            vads = vads_cache[owner_proc.vol.offset]
+        else:
+            vads = pe_symbols.PESymbols.get_proc_vads_with_file_paths(owner_proc)
+            vads_cache[owner_proc.vol.offset] = vads
+
+        # smear or terminated process
+        if len(vads) == 0:
+            return None
+
+        return vads
+
+    def _generator(
+        self,
+    ) -> Generator[
+        Tuple[
```

Comment:
The type checker definitely wants it, otherwise it complains that the return value isn't typed. We can typing: ignore generators if that is the plan, but I want it where eventually everything passes mypy checks

Reply:
I understand and I applaud the goal, so I guess we'll leave it in for now, but this does make the code uglier and does come with a comprehension cost even if it's somewhat mild.  I'm also impressed it hasn't run into an issue dealing with `ObjectInterfaces` and their `__getattr__` override so I'm not super hopeful for getting everything typed correctly (and I'd sooner accept some errors than paper over them with ignore command).

Reply:
You can reduce it a bit changing the Generator[Tuple[], None, None] for Iterable[Tuple[]] .. see https://github.com/volatilityfoundation/volatility3/blob/d56cd83510e64f9f875ff1dad11d8c8cbf5995f5/volatility3/framework/plugins/linux/capabilities.py#L162

Reply:
Yeah, weirdly I don't mind the generators too much, it's mostly when the generator returns a tuple of 15 types and we list each and every one of them...  The feels a bit of a waste, particularly since it's duplicated and checked as part of the actual running code...

Reply:
Can I leave this then or does it need to be changed still?

Reply:
I'll leave it up to you, I think it's a bit of an eyesore on the code, but if you think it's important for the typing then you can keep it.  Just means more scrolling when looking through the code...

Reply:
Yes I prefer it so the type checker is happy